text/3816-bidirectional-hash-maps.md

```diff
@@ -0,0 +1,108 @@
+- Feature Name: (`bidirectional_hash_maps`)
+- Start Date: (2025-05-21)
+- RFC PR: [rust-lang/rfcs#3816](https://github.com/rust-lang/rfcs/pull/3816)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+This RFC suggests a new data structure for Rust, to be implemented later: A Bi-Directional HashMap, or BijectionMap<K, V, S> where both the key and value are hashable and can be looked up. 
+
+# Motivation
+Many times, people would wish for a simple reverse-lookup function to convert their enums to different values, perhaps a String to their enum values and vice versa. In the current implementation, you would require two hashmaps:
+
+```rust
+use std::collections::HashMap;
+
+pub enum MyEnum {
+    VariantA,
+    VariantB,
+    VariantC,
+    VariantD,
+    // and so on...
+}
+
+const STRING_TO_MYENUM_MAP: HashMap<&str, MyEnum> = HashMap::from([("a", MyEnum::VariantA), ("b", MyEnum::VariantB), ("c", MyEnum::VariantC), ("d", MyEnum::VariantD)] /* and so on... */);
+const MYENUM_TO_STRING_MAP: HashMap<MyEnum, &str> = HashMap::from([(MyEnum::VariantA, "a"), (MyEnum::VariantB, "b"), (MyEnum::VariantC, "c"), (MyEnum::VariantD, "d")] /* and so on...*/);
+This is too cumbersome for many people to use. True, they could implement their own method, but could there be a better solution using just one map? They could define it
+
+use std::collections::BijectionMap;
+
+pub enum MyEnum {
+    VariantA,
+    VariantB,
+    VariantC,
+    VariantD,
+}
+
+const STRING_CONV_MYENUM_MAP: BijectionMap<&str, MyEnum> = BijectionMap::from([("a", MyEnum::VariantA), ("b", MyEnum::VariantB), ("c", MyEnum::VariantC), ("d", MyEnum::VariantD)] /* and so on... */);
+```
+and use it here:
+
+```rust
+let from_string = STRING_CONV_MYENUM_MAP.get_value("a");
+let to_string = STRING_CONV_MYENUM_MAP.get_key(MyEnum::VariantA);
+```
+
+# Guide-Level Explanation
+A BijectionMap is a Rust collection that maps keys and values together. Unlike a HashMap, you can look up a BijectionMap both ways, by key or by value.
+
+Say you're developing a program and you need to check usernames and match them up with user emails and vice versa. You can, obviously, use two hashmaps and develop a reverser function. But why do two if you can use just one map?
+
+```rust
+let USER_BIDI_MAP = BijectionMap::new::<&str, &str<()
+for user in users {
+    USER_BIDI_MAP.insert(user.username, user.email);
+}
+let random_user_who_logged_in = "iamaverysmartperson";
+let their_email USER_BIDI_MAP.get(random_user_who_logged_in);
+It can also be used to convert enums to other values:
+
+pub enum IpResponse<Value, Error> {
```

Comment:
````suggestion
let their_email = USER_BIDI_MAP.get(random_user_who_logged_in);
```

It can also be used to convert enums to other values:

```rust
pub enum IpResponse<Value, Error> {
````